contracts/Staking.sol

```diff
@@ -0,0 +1,256 @@
+        }
+    }
+
+    /**
+     * @dev function allow user to withdraw sum of stake tokens
+     * by caller + collected reward for caller
+     *
+     * @notice if the user call this function before the end of staking period
+     * he will receive only 60% of his reward
+     */
+    function unStake() external override {
+        require(
+            investorList[msg.sender].stakingAmount > 0,
+            "Error: you are not investor"
+        );
+        _updateReward(msg.sender);
+        uint256 amount;
+        uint256 reward = investorList[msg.sender].reward;
+        if (block.timestamp < stakingParams.stakingFinishDate) {
+            reward = reward - ((reward * 40 ether) / ONE_HUNDRED_PERCENT);
+            amount = investorList[msg.sender].stakingAmount + reward;
+        } else {
+            amount = investorList[msg.sender].stakingAmount + reward;
+        }
+        stakingTotalAmount -= investorList[msg.sender].stakingAmount;
+        delete investorList[msg.sender].startDate;
+        investorList[msg.sender].finishDate = block.timestamp;
+        delete investorList[msg.sender].stakingAmount;
+        delete investorList[msg.sender].reward;
+        token.safeTransfer(msg.sender, amount);
+        emit UnStake(msg.sender, amount);
+    }
+
+    /**
+     * @dev internal function which work when someone
+     * call 'stake' or 'unstake' functions
+     *
+     * @notice Update list :
+     *
+     * Update 'rewardPerTokenStored' to actual value of collected reward
+     * Update 'lastTimeUpdate' - last time when some one stake or unstake
+     * Update 'investorList[caller_].reward' - how many tokens caller
+     * can withdraw like reward
+     * Update 'investorList[caller_].userRewardPerTokens' - how many reward
+     * user collect in last time when he stake or unstake
+     */
+    function _updateReward(address caller_) internal {
+        rewardPerTokenStored = _rewardPerToken();
+        lastTimeUpdate = block.timestamp;
+        investorList[caller_].reward = _earned(caller_);
+        investorList[caller_].userRewardPerTokens = rewardPerTokenStored;
+    }
+
+    /**
+     * @dev calculate how many reward collected by all users
+     */
+    function _rewardPerToken() internal view returns (uint256) {
+        uint256 rewardRate = ((stakingTotalAmount * apr) /
+            ONE_HUNDRED_PERCENT) / 365 days;
+        if (stakingTotalAmount == 0) {
+            return 0;
+        } else {
+            return
+                rewardPerTokenStored +
+                ((rewardRate * (block.timestamp - lastTimeUpdate) * 1e18)) /
+                stakingTotalAmount;
+        }
+    }
+
+    /**
+     * @dev calculate how many reward collected by each user
+     */
+    function _earned(address caller_) internal view returns (uint256) {
+        if (block.timestamp < stakingParams.stakingFinishDate) {
+            return
+                (investorList[caller_].stakingAmount *
+                    (_rewardPerToken() -
```

Comment:
just use rewardPerTokenStored not need call _rewardPerToken() for calculate again